arch/x86/adler32_avx512.c

```diff
@@ -8,9 +8,108 @@
 
 #ifdef X86_AVX512
 
-#include "adler32_avx512_tpl.h"
+#include "../../zbuild.h"
+#include "../../adler32_p.h"
+#include "../../adler32_fold.h"
+#include "../../cpu_features.h"
+#include <immintrin.h>
+#include "x86_intrins.h"
+#include "adler32_avx512_p.h"
 
-#define COPY
-#include "adler32_avx512_tpl.h"
+static inline uint32_t adler32_fold_copy_impl(uint32_t adler, uint8_t *dst, const uint8_t *src, size_t len, const int COPY) {
+    if (src == NULL) return 1L;
+    if (len == 0) return adler;
+
+    uint32_t adler0, adler1;
+    adler1 = (adler >> 16) & 0xffff;
+    adler0 = adler & 0xffff;
+
+rem_peel:
+    if (len < 64) {
+        /* This handles the remaining copies, just call normal adler checksum after this */
+        if (COPY) {
+            __mmask64 storemask = (0xFFFFFFFFFFFFFFFFUL >> (64 - len));
+            __m512i copy_vec = _mm512_maskz_loadu_epi8(storemask, src);
+            _mm512_mask_storeu_epi8(dst, storemask, copy_vec);
+        }
+
+#ifdef X86_AVX2
+        return adler32_avx2(adler, src, len);
+#elif defined(X86_SSSE3)
+        return adler32_ssse3(adler, src, len);
+#else
+        return adler32_len_16(adler0, src, len, adler1);
+#endif
+    }
+
+    __m512i vbuf, vs1_0, vs3;
+
+    const __m512i dot2v = _mm512_set_epi8(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
+                                          20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37,
+                                          38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55,
+                                          56, 57, 58, 59, 60, 61, 62, 63, 64);
+    const __m512i dot3v = _mm512_set1_epi16(1);
+    const __m512i zero = _mm512_setzero_si512();
+    size_t k;
+
+    while (len >= 64) {
+        __m512i vs1 = _mm512_zextsi128_si512(_mm_cvtsi32_si128(adler0));
+        __m512i vs2 = _mm512_zextsi128_si512(_mm_cvtsi32_si128(adler1));
+        vs1_0 = vs1;
```

Comment:
This gets optimized away as it is only referenced inside the while loop starting from line 65 below. Moving the line to start of while loop might cause it to not get optimized away as long as line 86 is also removed. Codecov essentially doesn't know which one of these two lines are still in the resulting code after optimizer combines the two lines.

Reply:
I'm not entirely sure what you're proposing here.  We definitely want this assignment outside of the inner while loop.  Are we that worried about codecov for this tight loop?

Reply:
A lot of times codecov is correct about dead code. It isn't however correct in 100% of the cases, so I usually trace things on paper to see if a line gets optimized away. In this case the line you want to be outside the inner loop gets moved inside due to duplicate code elimination.

Reply:
I think this PR should only be able the refactoring. I don't suggest adding any functional or optimization changes. To me it looks like that code was already there.